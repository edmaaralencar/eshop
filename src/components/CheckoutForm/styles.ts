import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    padding: 2.4rem;
    background-color: #f5f8fa;

    h2 {
      color: ${theme.colors.title};
      font-size: 2rem;
      margin-bottom: 2rem;
    }
  `}
`

export const ButtonContainer = styled.div`
  margin-top: 2.4rem;

  display: grid;
  grid-template-columns: 1fr 1fr;
  gap: 0.8rem;
  align-items: center;

  a {
    text-decoration: none;
    text-align: center;
    font-size: 1.8rem;
    font-weight: 300;

    color: ${({ theme }) => theme.colors.primary.main};

    transition: color 0.1s ease-in;

    &:hover {
      color: ${({ theme }) => theme.colors.primary.dark};
    }
  }
`

export const Error = styled.p`
  color: red;
  font-size: 1.4rem;
  margin: 1.4rem 0;
`
