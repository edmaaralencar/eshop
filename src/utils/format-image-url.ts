export function formatImageUrl(image: string) {
  return `/products/${image}`
}
