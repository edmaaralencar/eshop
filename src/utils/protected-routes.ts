import { GetServerSidePropsContext } from 'next'
// import { unstable_getServerSession } from 'next-auth'
import { getSession } from 'next-auth/react'
// import { options } from 'pages/api/auth/[...nextauth].api'

async function protectedRoutes(context: GetServerSidePropsContext) {
  // const session = await unstable_getServerSession(
  //   context.req,
  //   context.res,
  //   options
  // )

  // if (!session) {
  //   return {
  //     redirect: {
  //       destination: `/sign-in?callbackUrl=${context.resolvedUrl}`,
  //       permanent: false
  //     }
  //   }
  // }

  const session = await getSession(context)

  if (!session) {
    context.res.setHeader(
      'Location',
      `/sign-in?callbackUrl=${context.resolvedUrl}`
    )
    context.res.statusCode = 302
  }

  return session
}

export default protectedRoutes
