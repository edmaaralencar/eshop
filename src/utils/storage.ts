const APP_KEY = '@eshop'

export function getStorageItem(key: string) {
  if (typeof window === 'undefined') return

  const data = window.localStorage.getItem(`${APP_KEY}.${key}`)
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return JSON.parse(data!)
}

export function setStorageItem(key: string, value: any[]) {
  if (typeof window === 'undefined') return

  const data = JSON.stringify(value)
  return window.localStorage.setItem(`${APP_KEY}.${key}`, data)
}

export function removeStorageItem(key: string) {
  if (typeof window === 'undefined') return

  return window.localStorage.removeItem(`${APP_KEY}.${key}`)
}
