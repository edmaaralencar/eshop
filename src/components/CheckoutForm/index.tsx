import Link from 'next/link'
import { useState } from 'react'
import { PaymentElement, useElements, useStripe } from '@stripe/react-stripe-js'
import { useRouter } from 'next/router'

import { useCart } from 'hooks/use-cart'
import { api } from 'lib/axios'

import Button from 'components/Button'

import * as S from './styles'

// function CheckoutForm({ error, setError }: CheckoutFormProps) {
function CheckoutForm() {
  const [error, setError] = useState<string | null>(null)
  const { clearCart, cartItems } = useCart()
  const stripe = useStripe()
  const elements = useElements()
  const router = useRouter()

  async function handleSubmit(event: React.FormEvent) {
    event.preventDefault()

    if (!stripe || !elements) {
      return
    }

    const payload = await stripe.confirmPayment({
      elements,
      redirect: 'if_required'
    })

    if (payload.error) {
      // setError(`Payment failed ${payload.error.message}`)
      setError(String(payload.error.message))
    } else {
      setError(null)

      clearCart()

      const { data } = await api({
        method: 'POST',
        url: '/orders',
        data: {
          cartItems,
          paymentIntentId: payload.paymentIntent?.id,
          paymentMethod: payload?.paymentIntent?.payment_method,
          total: payload.paymentIntent?.amount
        }
      })

      await router.push(`/success?id=${data.id}`)
    }
  }

  return (
    <S.Wrapper>
      <h2>Pagamento</h2>
      <form onSubmit={handleSubmit}>
        <PaymentElement />

        {error && <S.Error>{error} Tente novamente.</S.Error>}

        <S.ButtonContainer>
          <Link href="/products">
            <a href="">Continuar comprando</a>
          </Link>
          <Button type="submit" fullWidth disabled={!stripe || !elements}>
            Pagar
          </Button>
        </S.ButtonContainer>
      </form>
    </S.Wrapper>
  )
}

export default CheckoutForm
