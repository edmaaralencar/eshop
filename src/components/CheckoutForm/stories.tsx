import { StoryObj, Meta } from '@storybook/react'
import CheckoutForm from '.'

export default {
  title: 'CheckoutForm',
  component: CheckoutForm
} as Meta

export const Default: StoryObj = {}
