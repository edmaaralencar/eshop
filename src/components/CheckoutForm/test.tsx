import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test-helper'

import CheckoutForm from '.'

jest.mock('@stripe/react-stripe-js', () => ({
  PaymentElement: function Mock({ children }: { children: React.ReactNode }) {
    return <div data-testid="Mock PaymentElement">{children}</div>
  },
  Elements: function Mock({ children }: { children: React.ReactNode }) {
    return <div data-testid="Mock Elements">{children}</div>
  },
  useStripe: jest.fn().mockReturnValue({
    confirmPayment: jest.fn().mockResolvedValue({
      elements: ''
    })
  }),
  useElements: jest.fn().mockReturnValue({
    getElement: jest.fn()
  })
}))

// eslint-disable-next-line @typescript-eslint/no-var-requires
const useRouter = jest.spyOn(require('next/router'), 'useRouter')
const push = jest.fn()

useRouter.mockImplementation(() => ({
  push,
  query: '',
  asPath: '',
  route: '/'
}))

describe('<CheckoutForm />', () => {
  it('should render the component correctly', () => {
    const { container } = renderWithTheme(<CheckoutForm />)

    expect(
      screen.getByRole('heading', { name: /pagamento/i })
    ).toBeInTheDocument()

    expect(screen.getByTestId(/Mock PaymentElement/i)).toBeInTheDocument()

    expect(screen.getByRole('button', { name: /pagar/i })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
