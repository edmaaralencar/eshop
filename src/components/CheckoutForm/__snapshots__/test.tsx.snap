// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckoutForm /> should render the component correctly 1`] = `
.c2 {
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #6C6CFF;
  color: #FFFFFF;
  -webkit-transition: background-color 0.2s ease-out;
  transition: background-color 0.2s ease-out;
  padding: 0 2rem;
  height: 5.2rem;
  width: 100%;
}

.c2 span {
  font-weight: 300;
}

.c2:disabled {
  cursor: not-allowed;
}

.c2:hover {
  background-color: #7676ff;
}

.c2 span {
  font-size: 1.8rem;
}

.c0 {
  padding: 2.4rem;
  background-color: #f5f8fa;
}

.c0 h2 {
  color: #18183F;
  font-size: 2rem;
  margin-bottom: 2rem;
}

.c1 {
  margin-top: 2.4rem;
  display: grid;
  grid-template-columns: 1fr 1fr;
  gap: 0.8rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 a {
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  font-size: 1.8rem;
  font-weight: 300;
  color: #8787FF;
  -webkit-transition: color 0.1s ease-in;
  transition: color 0.1s ease-in;
}

.c1 a:hover {
  color: #6C6CFF;
}

<div
  class="c0"
>
  <h2>
    Pagamento
  </h2>
  <form>
    <div
      data-testid="Mock PaymentElement"
    />
    <div
      class="c1"
    >
      <a
        href="/products"
      >
        Continuar comprando
      </a>
      <button
        class="c2"
        type="submit"
      >
        <span>
          Pagar
        </span>
      </button>
    </div>
  </form>
</div>
`;
